core/image/etc/etc2_compressor.cpp

```diff
@@ -0,0 +1,148 @@
+    default:
+      return Etc::Image::Format::UNKNOWN;
+  }
+}
+
+void read_image(const uint8_t* input_image, uint32_t width, uint32_t height,
+                Etc::ColorFloatRGBA* output) {
+  const uint8_t BYTE_PER_PIXEL = 4;
+  for (uint32_t h = 0; h < height; ++h) {
+    const uint8_t* src = &input_image[(h * width) * BYTE_PER_PIXEL];
+    Etc::ColorFloatRGBA* dst = &output[h * width];
+    for (uint32_t w = 0; w < width; ++w) {
+      *dst =
+          Etc::ColorFloatRGBA::ConvertFromRGBA8(src[0], src[1], src[2], src[3]);
+      dst++;
+      src += BYTE_PER_PIXEL;
+    }
+  }
+}
+
+extern "C" etc_error compress_etc(const uint8_t* input_image,
+                                  uint8_t* output_image, uint32_t width,
+                                  uint32_t height, etc_format format) {
+  auto* source_image = new Etc::ColorFloatRGBA[width * height];
+  if (!source_image) {
+    return ETC_ALLOCATION_FAILED;
+  }
+  read_image(input_image, width, height, source_image);
+  Etc::Image image(reinterpret_cast<float*>(source_image), width, height,
+                   ERROR_METRIC);
+  image.m_bVerboseOutput = false;
+
+  auto image_format = convert_etc_format(format);
+  if (image_format == Etc::Image::Format::UNKNOWN) {
+    return static_cast<etc_error>(
+        Etc::Image::EncodingStatus::ERROR_UNKNOWN_FORMAT);
+  }
+
+  auto status =
+      image.Encode(image_format, ERROR_METRIC, EFFORT, MIN_JOBS, MAX_JOBS);
+  // We don't need to care about warnings.
+  if (status > Etc::Image::EncodingStatus::ERROR_THRESHOLD) {
+    delete[] source_image;
+    return static_cast<etc_error>(status);
+  }
+
+  memcpy(output_image, image.GetEncodingBits(), image.GetEncodingBitsBytes());
+  delete[] source_image;
+  return static_cast<etc_error>(Etc::Image::EncodingStatus::SUCCESS);
+}
+
+extern "C" char* get_etc_error_string(etc_error error_code) {
+  char* error_string = (char*)calloc(512, sizeof(char));
+  if (error_code == ETC_ALLOCATION_FAILED) {
+    strcpy(error_string, "Allocation Failed");
+    return error_string;
+  }
+
+  auto status = static_cast<Etc::Image::EncodingStatus>(error_code);
+  if (status == Etc::Image::EncodingStatus::SUCCESS) {
+    strcpy(error_string, "Compression Succeed");
+    return error_string;
+  }
+
+  char* current = error_string;
+  int written = sprintf(current, "[");
+  current += written;
+
+  if (status > Etc::Image::EncodingStatus::ERROR_THRESHOLD) {
+    if (status & Etc::Image::EncodingStatus::ERROR_ZERO_WIDTH_OR_HEIGHT) {
+      written = sprintf(current, "\"Error: Image width or height is zero\"");
+      current += written;
+    }
+    if (status & Etc::Image::EncodingStatus::ERROR_ZERO_WIDTH_OR_HEIGHT) {
+      written = sprintf(current, "\"Error: Image width or height is zero\"");
+      current += written;
+    }
+    if (status & Etc::Image::EncodingStatus::ERROR_ZERO_WIDTH_OR_HEIGHT) {
+      written = sprintf(current, "\"Error: Image width or height is zero\"");
+      current += written;
```

Comment:
Three times the same check and message?

I guess here you want to collect all possible errors from the Etc::Image::EncodingStatus bitfield, by allocating a string on the head and `sprintf()` into it. However, I guess callers of get_etc_error_string will never free this string, leading to a memory leak?

For the sake of simplicity, I would just return an error message for the first error bit encountered, e.g.:

```
if (status & Etc::Image::EncodingStatus::ERROR_ZERO_WIDTH_OR_HEIGHT) {
  return "Zero width of height";
}
// same with other error bits
```

I think it's OK to list one error as this is a fail-path anyway, the subsequent logic of the program won't depend on the string you return here.

Reply:
So as we have to return a const pointer here, it's really between the choice of "returning multiple errors" so we can log them all so in case of any bug, it would be helpful to us or prevent a memory leak. I have chosen the multiple error.

My rationing is, this method will never be called in a well behaving path. When this method is called, it means that something is wrong and the experiment will fail and we will show user an error and probably user will shut down/restart. In a 5GB memory user application, leaking half a KB is not a big problem(as the application will be closed soon anyway). Therefore, I believe having all errors logged is more important and gives more benefit to both user and us than preventing the memory leak here. 

If you strongly oppose, we can change it or, try to find a middle ground(like on multiple call, check if error string is allocated and if allocated then free, so we limit it with 512 bytes) etc. But I don't think that it will worth the effort.

Reply:
I'm ok with the memory leak if there's a small comment to acknowledge it, such that when we see this code in a few months we understand the choice that was made, e.g.:

```
// Heap-alloc a small string to format the error message. This is a
// small memory leak that we deem acceptable since the program
// is likely to be closed after such an error.
```